transaction/std/src/memo_builder/gift_code_funding_memo_builder.rs

```diff
@@ -0,0 +1,426 @@
+    gift_code_tx_out_public_key: Option<RistrettoPublic>,
+    // Whether or not to enable change memo
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+    // Whether we've written an unused memo to the destination
+    wrote_destination_memo: bool,
+    // Whether our last note setting attempt was invalid and the note hasn't
+    // been reset with a valid note or cleared
+    attempted_invalid_note: bool,
+}
+
+// Create an empty GiftCodeFundingMemoBuilder
+impl Default for GiftCodeFundingMemoBuilder {
+    fn default() -> Self {
+        Self {
+            note: "".into(),
+            gift_code_tx_out_public_key: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+            wrote_destination_memo: false,
+            attempted_invalid_note: false,
+        }
+    }
+}
+
+impl GiftCodeFundingMemoBuilder {
+    /// Set a utf-8 note (up to 60 bytes) onto the funding memo indicating
+    /// what the gift code was for. This method will enforce the 60 byte
+    /// limit with a NewMemoErr if the &str passed is longer than
+    /// 60 bytes.
+    pub fn set_funding_note(&mut self, note: &str) -> Result<(), NewMemoError> {
+        if note.len() > GiftCodeFundingMemo::NOTE_DATA_LEN {
+            self.attempted_invalid_note = true;
+            return Err(NewMemoError::BadInputs(
+                "Note memo cannot be greater than 60 bytes".into(),
+            ));
+        }
+        self.attempted_invalid_note = false;
+        self.note = note.into();
+        Ok(())
+    }
+    /// Clear the gift code funding note
+    pub fn clear_funding_note(&mut self) {
+        self.attempted_invalid_note = false;
+        self.note = "".into();
+    }
+    /// Set the TxOut public_key of the gift code TxOut sent to the
+    /// reserved gift code subaddress.
+    ///
+    /// IMPORTANT NOTE: Do NOT pass the public_key of the change TxOut
+    /// that the gift code memo is attached to as an argument. Doing
+    /// so will result in an error when attempting to build the memo.
+    pub fn set_gift_code_tx_out_public_key(
+        &mut self,
+        tx_out_public_key: &RistrettoPublic,
+    ) -> Result<(), NewMemoError> {
+        self.gift_code_tx_out_public_key = Some(*tx_out_public_key);
+        Ok(())
+    }
+    /// Clear the gift code tx_out_public_key
+    pub fn clear_gift_code_tx_out_public_key(&mut self) {
+        self.gift_code_tx_out_public_key = None;
+    }
+    /// Enable change memos
+    pub fn enable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = true;
+    }
+    /// Disable change memos
+    pub fn disable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = false;
+    }
+}
+
+impl MemoBuilder for GiftCodeFundingMemoBuilder {
+    /// Set the fee
+    fn set_fee(&mut self, _fee: Amount) -> Result<(), NewMemoError> {
+        Ok(())
```

Comment:
After checking with @garbageslam we think that we should actually track the fee being paid in the funding memo.

Reply:
Sure. At this point though I'm inclined to make that a change to the MCIP and allow people a day to make comments on it because that puts gift codes into RTH territory & takes a way a decent amount of space from the memo - and given Moby has driven a lot of this, that change is something we should have other teams who will be using it (mainly Moby/mobile teams) weigh in on.

I'd like to address that in a separate PR scoped to adding information to the memos.

Reply:
We discussed this over Slack and it will be addressed in a followup PR.